### PR TITLE
docs: Jest matchers docs

### PR DESCRIPTION
**Pull request by mdjastrzebski**
### Summary

Create API docs & migration guide for built-in Jest matchers.

This will not be made public yet, until all of the matchers are migrated and verified. 

### Test plan

Read the docs.

### CODE REVIEW

File: website/docs/JestMatchers.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+This allows you to assert whether the given element has specified accessible value.
+
+## `toBeEnabled()` / `toBeDisabled`
+
+```ts
+expect(element).toBeEnabled()
+expect(element).toBeDisabled()
+```
+
+These allows you to assert whether the given element is enabled or disabled from user's perspective. It relies on accessibility disabled state as set by `aria-disabled` or `accessibilityState.disabled` props. It will considers given element disabled when it or any of its ancestors is disabled.
```
Comment (pierrezimmermannbam) [inline website/docs/JestMatchers.md:111]: It's rather if any parent than any ancestor no?

File: website/docs/JestMatchers.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+:::note
+Built-in Jest matchers require RNTL v12.4.0 or later.
+:::
+
+This guide describes built-in Jest matchers, we recommend using these matchers as they provide more readable tests, better accessibility support and better developer experience.
+
+If you are already using legacy Jest Native matchers we have a [migration guide](migration-jest-native) for moving to the built-in matchers.
+
+import TOCInline from '@theme/TOCInline';
```
Comment (pierrezimmermannbam) [inline website/docs/JestMatchers.md:14]: What does this mean?

File: website/docs/JestMatchers.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+```
+
+This allows you to assert whether element is attached to the element tree or not. If you hold a reference to an element and it gets unmounted during the test it will no longer pass this assertion.
+
+## Element Content
+
+### `toHaveTextContent()`
+
+```ts
+expect(element).toHaveTextContent(t
```
Comment (pierrezimmermannbam) [inline website/docs/JestMatchers.md:33]:
```suggestion
expect(element).toHaveTextContent(
```

File: website/docs/JestMatchers.md [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+  options?: {
+    exact?: boolean;
+    normalizer?: (text: string) => string;
+  },
+)
+```
+
+This allows you to assert whether the given element has a text content or not. It accepts either `string` or `RegExp` matchers, as well as [text match options](Queries.md#text-match-options) of `exact` and `normalizer`.
+
+When `text` parameter is `undefined` it will only check for existence of text content, and when `text` is defined it will check if the actual text content matches passed value.
```
Comment (pierrezimmermannbam) [inline website/docs/JestMatchers.md:44]: Based on the signature It does not look like it can be unefined?

File: website/docs/JestMatchers.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+```ts
+expect(element).toBeEnabled()
+expect(element).toBeDisabled()
+```
+
+These allows you to assert whether the given element is enabled or disabled from user's perspective. It relies on accessibility disabled state as set by `aria-disabled` or `accessibilityState.disabled` props. It will considers given element disabled when it or any of its ancestors is disabled.
+
+:::note
+This matchers are negation of each other, and both are probivided to avoid double negations in your assertions.
```
Comment (pierrezimmermannbam) [inline website/docs/JestMatchers.md:114]:
```suggestion
This matchers are negation of each other, and both are provided to avoid double negations in your assertions.
```

File: website/docs/JestMatchers.md [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+```ts
+expect(element).toBeChecked()
+expect(element).toBePartiallyChecked()
+```
+
+These allows you to assert whether the given element is checked or partially checked from user's perspective. It relies on accessibility checked state as set by `aria-checked` or `accessibilityState.checked` props.
+
+:::note
+* `toBeChecked()` matcher works only on elements with `checkbox` or `radio` role.
+* `toBePartiallyChecked()` matchers works only on elements with `checkbox` role.
```
Comment (pierrezimmermannbam) [inline website/docs/JestMatchers.md:138]:
```suggestion
* `toBePartiallyChecked()` matcher works only on elements with `checkbox` role.
```

Comment (pierrezimmermannbam) [reply website/docs/JestMatchers.md:138]: What is the difference between the two except the constraint on the role?

File: website/docs/JestMatchers.md [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+```ts
+expect(element).toBeExpanded()
+expect(element).toBeCollapsed()
+```
+
+These allows you to assert whether the given element is expanded or collapsed from user's perspective. It relies on accessibility disabled state as set by `aria-expanded` or `accessibilityState.expanded` props.
+
+:::note
+This matchers are negation of each other for expandable elements (elements with explicit `aria-expanded` or `accessibilityState.expanded` props). However, both won't pass for non-expandable elements (ones without explicit `aria-expanded` or `accessibilityState.expanded` props).
```
Comment (pierrezimmermannbam) [inline website/docs/JestMatchers.md:151]:
```suggestion
These matchers are negation of each other for expandable elements (elements with explicit `aria-expanded` or `accessibilityState.expanded` props). However, both won't pass for non-expandable elements (ones without explicit `aria-expanded` or `accessibilityState.expanded` props).
```

File: website/docs/JestMatchers.md [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+### `toBeVisible()`
+
+```ts
+expect(element).toBeVisible()
+```
+
+This allows you to assert whether the given element is visible from user's perspective. 
+
+The element is considered invisibile when it or any of its ancestors has `display: none` or `opacity: 0` styles, as well as when it's hidden from accessbility.
```
Comment (pierrezimmermannbam) [inline website/docs/JestMatchers.md:173]:
```suggestion
The element is considered invisible when itself or any of its parents has `display: none` or `opacity: 0` styles, as well as when it's hidden from accessbility.
```

File: website/docs/JestMatchers.md [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+expect(element).toHaveProp(
+  name: string,
+  value?: unknown,
+)
+```
+
+This allows you to assert whether the given element has a given prop. When `value` parameter is `undefined` it will only check for existence of prop, and when `value` is defined it will check if the actual value matches passed value.
+
+:::note
+This matchers should be treated as escape hatch to be used when all other matchers are not suitable.
```
Comment (pierrezimmermannbam) [inline website/docs/JestMatchers.md:217]:
```suggestion
This matcher should be treated as escape hatch to be used when all other matchers are not suitable.
```

File: website/docs/MigrationJestMatchers.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
+---
+id: migration-jest-native
+title: Migration from Jest Native matchers
+---
+
+This guide describes steps necessary to migrate from [legacy Jest Native matchers v5](https://github.com/testing-library/jest-native) to [built-in Jest matchers](jest-matchers).
```
Comment (pierrezimmermannbam) [inline website/docs/MigrationJestMatchers.md:6]:
```suggestion
This guide describes the necessary steps to migrate from [legacy Jest Native matchers v5](https://github.com/testing-library/jest-native) to [built-in Jest matchers](jest-matchers). 
```

File: website/docs/MigrationJestMatchers.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
+---
+id: migration-jest-native
+title: Migration from Jest Native matchers
+---
+
+This guide describes steps necessary to migrate from [legacy Jest Native matchers v5](https://github.com/testing-library/jest-native) to [built-in Jest matchers](jest-matchers). 
+
+import TOCInline from '@theme/TOCInline';
```
Comment (pierrezimmermannbam) [inline website/docs/MigrationJestMatchers.md:8]: Should be removed?

File: website/docs/MigrationJestMatchers.md (hunk 10 above)
Comment (pierrezimmermannbam) [inline website/docs/MigrationJestMatchers.md:6]:
```suggestion
This guide describes the steps necessary to migrate from [legacy Jest Native matchers v5](https://github.com/testing-library/jest-native) to [built-in Jest matchers](jest-matchers). 
```

File: website/docs/MigrationJestMatchers.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+### Added matchers
+
+New [`toHaveAccessibleName()`](jest-matchers#tohaveaccessiblename) has been added.
+
+### Noteworthy details
+
+You should be aware of following changes:
+* [`toBeEnabled()` / `toBeDisabled()`](jest-matchers#tobeenabled) matchers also check the disabled state for element ancestors and not only the element itself
+* [`toBeChecked()`](jest-matchers#tobechecked) matcher supports only elements with `checkbox` or `radio` role
+* [`toBePartiallyChecked()`](jest-matchers#tobechecked) matchers supports only elements with `checkbox` role
```
Comment (pierrezimmermannbam) [inline website/docs/MigrationJestMatchers.md:79]:
```suggestion
* [`toBePartiallyChecked()`](jest-matchers#tobechecked) matcher supports only elements with `checkbox` role
```

File: website/docs/MigrationJestMatchers.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+The new matchers support both `accessbililityState` and `aria-*` props.
+
+### Added matchers
+
+New [`toHaveAccessibleName()`](jest-matchers#tohaveaccessiblename) has been added.
+
+### Noteworthy details
+
+You should be aware of following changes:
+* [`toBeEnabled()` / `toBeDisabled()`](jest-matchers#tobeenabled) matchers also check the disabled state for element ancestors and not only the element itself
```
Comment (pierrezimmermannbam) [inline website/docs/MigrationJestMatchers.md:77]: That was already done by jest native I believe https://github.com/testing-library/jest-native/blob/main/src/to-be-disabled.ts#L36

**Review by pierrezimmermannbam — CHANGES_REQUESTED** [review]
Very thorough documentation! I left some minor comments (mostly typos) to be fixed and it will be perfect

File: website/docs/JestMatchers.md (hunk 6 above)
Comment (mdjastrzebski) [reply website/docs/JestMatchers.md:138]: `toBeChecked` matches `checked: true:` state, while `toBePartiallyChecked` matches `checked: "mixed'`

File: website/docs/JestMatchers.md (hunk 4 above)
Comment (mdjastrzebski) [reply website/docs/JestMatchers.md:44]: Will check, thanks for catching this.

File: website/docs/MigrationJestMatchers.md (hunk 13 above)
Comment (mdjastrzebski) [reply website/docs/MigrationJestMatchers.md:77]: True, I'll rephrase this to highlight the migration change from `toHaveAccessibilityState({ disabled: true })` which was not examining ancestors.

**Comment by mdjastrzebski** [comment]
@pierrezimmermannbam looking at the number of spelling mistakes I need to invest into some spell checker (maybe ChatGPT 😉). Thanks for catching these!

File: website/docs/JestMatchers.md (hunk 4 above)
Comment (mdjastrzebski) [reply website/docs/JestMatchers.md:44]: Good catch, dropping the "undefined" version.

File: website/docs/MigrationJestMatchers.md (hunk 13 above)
Comment (mdjastrzebski) [reply website/docs/MigrationJestMatchers.md:77]: Rephrased to make more sense.

**Review by pierrezimmermannbam — APPROVED** [review]
Well it looks like I still missed a lot, I guess gpt is better than I am at this. Anyways this looks good to me, great work!
